plugins/blockvolume/blockprovider/gluster-block/glusterblock.go

```diff
@@ -0,0 +1,274 @@
+package glusterblock
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/pkg/size"
+	"github.com/gluster/glusterd2/plugins/blockvolume/blockprovider"
+	"github.com/gluster/glusterd2/plugins/blockvolume/utils"
+
+	"github.com/gluster/gluster-block-restapi/client"
+	"github.com/gluster/gluster-block-restapi/pkg/api"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+const providerName = "gluster-block"
+
+func init() {
+	log.WithField("name", providerName).Infof("Registering block provider")
+	blockprovider.RegisterBlockProvider(providerName, newGlusterBlock)
+}
+
+// GlusterBlock implements block Provider interface. It represents a gluster-block
+type GlusterBlock struct {
+	client     client.GlusterBlockClient
+	ClientConf *ClientConfig
+}
+
+func newGlusterBlock() (blockprovider.Provider, error) {
+	var (
+		gb         = &GlusterBlock{}
+		clientConf = &ClientConfig{}
+		opts       = []client.OptFuncs{}
+	)
+
+	clientConf.ApplyFromConfig(viper.GetViper())
+	gb.ClientConf = clientConf
+
+	opts = append(opts,
+		client.WithAuth(clientConf.User, clientConf.Secret),
+		client.WithTLSConfig(&client.TLSOptions{CaCertFile: clientConf.CaCertFile, InsecureSkipVerify: clientConf.Insecure}),
+	)
+
+	gbClient, err := client.NewClientWithOpts(clientConf.HostAddress, opts...)
+	if err != nil {
+		return nil, err
+	}
+	gb.client = gbClient
+
+	return gb, nil
+}
+
+// CreateBlockVolume will create a gluster block volume with given name and size having `hostVolume` as hosting volume
+func (g *GlusterBlock) CreateBlockVolume(name string, size uint64, hosts []string, hostVolume string, options ...blockprovider.BlockVolOption) (blockprovider.BlockVolume, error) {
+	var (
+		blockVolOpts = &blockprovider.BlockVolumeOptions{}
+		clusterLocks = transaction.Locks{}
+	)
+
+	blockVolOpts.ApplyOpts(options...)
+
+	req := &api.BlockVolumeCreateReq{
+		HaCount:            blockVolOpts.Ha,
+		AuthEnabled:        blockVolOpts.Auth,
+		FullPrealloc:       blockVolOpts.FullPrealloc,
+		Size:               size,
+		Storage:            blockVolOpts.Storage,
+		RingBufferSizeInMB: blockVolOpts.RingBufferSizeInMB,
+		Hosts:              hosts,
+	}
+
+	if err := clusterLocks.Lock(hostVolume); err != nil {
+		log.WithError(err).Error("error in acquiring cluster lock")
+		return nil, err
+	}
+	defer clusterLocks.UnLock(context.Background())
+
+	volInfo, err := volume.GetVolume(hostVolume)
```

Comment:
can we move this logic to some other function, so this can be reused for checking block hosting volume?

Reply:
done